examples/tox21/keras_gc.py

```diff
@@ -0,0 +1,253 @@
+    for deg in range(1, self.max_degree + 1):
+      gathered_atoms = tf.gather(atoms, deg_adj_lists[deg - 1])
+      # Sum along neighbors as well as self, and store
+      summed_atoms = tf.reduce_sum(gathered_atoms, 1)
+      deg_summed[deg - 1] = summed_atoms
+
+    return deg_summed
+
+
+class GraphGather(tf.keras.layers.Layer):
+
+  def __init__(self, batch_size, activation_fn, **kwargs):
+    self.batch_size = batch_size
+    self.activation_fn = activation_fn
+    super(GraphGather, self).__init__(**kwargs)
+
+  def call(self, x, **kwargs):
+    atom_features = x[0]
+
+    # Extract graph topology
+    membership = x[1]
+    activated_par = tf.dynamic_partition(atom_features, membership,
+                                         self.batch_size)
+
+    # Sum over atoms for each molecule
+    sparse_reps = [
+        tf.reduce_mean(activated, 0, keep_dims=True)
+        for activated in activated_par
+    ]
+    max_reps = [
+        tf.reduce_max(activated, 0, keep_dims=True)
+        for activated in activated_par
+    ]
+
+    # Get the final sparse representations
+    sparse_reps = tf.concat(axis=0, values=sparse_reps)
+    max_reps = tf.concat(axis=0, values=max_reps)
+    mol_features = tf.concat(axis=1, values=[sparse_reps, max_reps])
+
+    if self.activation_fn is not None:
+      mol_features = self.activation_fn(mol_features)
+    out_tensor = mol_features
+    return out_tensor
+
+  def compute_output_shape(self, input_shape):
+    return (self.batch_size, 2 * input_shape[0][-1])
+
+
+batch_size = 64
+atom_features = tf.placeholder(
+    tf.float32, shape=(
+        None,
+        75,
+    ))
+degree_slice = tf.placeholder(tf.int32, shape=(None, 2))
+membership = tf.placeholder(tf.int32, shape=(None,))
+
+deg_adjs = []
+for i in range(10):
+  deg_adj = tf.placeholder(
+      tf.int32, shape=(
+          None,
+          i + 1,
+      ))
+  deg_adjs.append(deg_adj)
+
+inputs = [atom_features, degree_slice, membership] + deg_adjs
+
+x = GraphConv(
+    64, activation_fn=tf.nn.relu)([atom_features, degree_slice] + deg_adjs)
+# TODO(LESWING) GraphPool Conversion
+x = tf.keras.layers.Dense(128)(x)
+x = tf.keras.layers.BatchNormalization()(x)
+readout = GraphGather(
+    batch_size=batch_size, activation_fn=tf.nn.tanh)([x, membership])
+
+labels = []
+weights = []
+losses = []
+outputs = []
+for i in range(len(tox21_tasks)):
+  logit = tf.keras.layers.Dense(2, activation=None)(readout)
+
+  label = tf.placeholder(tf.int32, shape=(None))
+  labels.append(label)
+
+  weight = tf.placeholder(tf.float32, shape=(None,))
+  weights.append(weight)
+
+  loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
+      labels=label, logits=logit) * weight
+  losses.append(loss)
+
+  output = tf.nn.softmax(logit)
+  outputs.append(output)
+
+final_loss = tf.reduce_mean(losses)
+train_step = tf.train.AdamOptimizer().minimize(final_loss)
+
+init_op = tf.global_variables_initializer()
+sess.run(init_op)
+
+with sess.as_default():
```

Comment:
I couldn't use Keras' Model interface because it does strict checking on your inputs and outputs.  For each batch each Input and Output have to have the same length in the zeroth dimension.  

This is impossible for the current featurization scheme of our graph models.  (And I can't think of any easy way to make any given batch have the same zeroth dimension for each input) @peastman @rbharath @miaecle .  We might be able to subclass Model and override the method to not do the strict checking but that could have really bad reprocussions elsewhere in the framework.

Reply:
Because we can't use the Model class this code gets a lot more ugly really fast.  We still need the tight coupling between the dataset and the Model from TensorGraph.

Reply:
Are you sure about that?  I'm looking through the code for `Model.fit_generator()` and I don't see any restriction like that.  On the contrary, for every batch it separately looks up the batch size and logs it: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/_impl/keras/engine/training.py#L2160-L2169.  Can you post a simple example that shows the problem?

Reply:
I was looking at fit not fit_generator().  Give me some time to rebundle the keras_gc more in keras land (replace tf.placeholder with tf.keras.layers.Input, etc) and see if we can use fit_generator().

Reply:
`fit()` looks like a more limited function.  You give it numpy arrays containing the entire dataset, and it breaks them up into batches for you.  `fit_generator()` lets you decide how to break them into batches.

Reply:
it is still happening.  Check out keras_gc_full_keras.py
```python
Epoch 1/1
Traceback (most recent call last):
  File "/home/leswing/Documents/deepchem/examples/tox21/keras_gc_full_keras.py", line 233, in <module>
    model.fit_generator(generator(train_dataset), 10)
  File "/home/leswing/miniconda3/envs/deepchem/lib/python3.5/site-packages/tensorflow/python/keras/_impl/keras/engine/training.py", line 2076, in fit_generator
    x, y, sample_weight=sample_weight, class_weight=class_weight)
  File "/home/leswing/miniconda3/envs/deepchem/lib/python3.5/site-packages/tensorflow/python/keras/_impl/keras/engine/training.py", line 1781, in train_on_batch
    check_batch_axis=True)
  File "/home/leswing/miniconda3/envs/deepchem/lib/python3.5/site-packages/tensorflow/python/keras/_impl/keras/engine/training.py", line 1410, in _standardize_user_data
    _check_array_lengths(x, y, sample_weights)
  File "/home/leswing/miniconda3/envs/deepchem/lib/python3.5/site-packages/tensorflow/python/keras/_impl/keras/engine/training.py", line 229, in _check_array_lengths
    [x.shape for x in inputs]))
ValueError: All input arrays (x) should have the same number of samples. Got array shapes: [(1279, 75), (11, 2), (1279, 1), (313, 1), (583, 2), (341, 3), (41, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10)]
```